src/cpp/runner/main.cpp

```diff
@@ -9,26 +10,144 @@
+{
+    std::ofstream out(filepath);
+    out << std::scientific << objective_time << "\t" << derivative_time;
+    out.close();
+}
+
+void save_objective_to_file(const string& filepath, const double value)
+{
+    std::ofstream out(filepath);
+    out << std::scientific << value;
+    out.close();
+}
+
+void save_gradient_to_file(const string& filepath, int gradient_size, double* gradient)
+{
+    std::ofstream out(filepath);
+
+    for (auto i = 0; i < gradient_size; i++)
+    {
+        out << std::scientific << gradient[i] << "\t";
+    }
+
+    out.close();
+}
 
-int main(int argc, const char* argv[])
+int main(const int argc, const char* argv[])
 {
-	if (argc < 7) {
-		std::cerr << "usage: CPPRunner modulePath dir_in dir_out file_basename nruns_F nruns_J [-rep]\n";
-		return 1;
-	}
+    try {
+        if (argc < 8) {
+            std::cerr << "usage: CPPRunner module_path input_filepath output_dir minimum_measurable_time nruns_F nruns_J time_limit [-rep]\n";
+            return 1;
+        }
+
+        const auto module_path = argv[1];
+        const string input_filepath(argv[2]);
+        const string output_dir(argv[3]);
+        const auto minimum_measurable_time = std::stod(argv[4]);
+        const auto nruns_F = std::stoi(argv[5]);
+        const auto nruns_J = std::stoi(argv[6]);
+        const auto time_limit = std::stod(argv[7]);
+
+        // read only 1 point and replicate it?
+        const auto replicate_point = (argc > 8 && string(argv[8]) == "-rep");
+
+        ModuleLoader module_loader(module_path);
+        auto test = module_loader.GetTest();
+
+        auto inputs = read_input_data(input_filepath, replicate_point);
+        auto gradient_size = (inputs.k * (inputs.d + 1) * (inputs.d + 2)) / 2;
+
+        test->prepare(std::move(inputs));
+
+        const auto objective_time =
+            measure_shortest_time(minimum_measurable_time, nruns_F, time_limit, test, &ITest<GMMInput, GMMOutput>::calculateObjective);
 
-	auto modulePath = argv[1];
-	string dir_in(argv[1]);
-	string dir_out(argv[2]);
+        const auto derivative_time =
+            measure_shortest_time(minimum_measurable_time, nruns_J, time_limit, test, &ITest<GMMInput, GMMOutput>::calculateJacobian);
 
-	ModuleLoader moduleLoader(modulePath);
-	auto test = moduleLoader.GetTest();
+        auto output = test->output();
 
-	/*// Read instance
-	auto inputs = readInputData();*/
+        auto input_basename = filepath_to_basename(input_filepath);
+        auto module_basename = filepath_to_basename(module_path);
 
-	test->output();
-}
\ No newline at end of file
+        save_time_to_file(output_dir + input_basename + "_times_" + module_basename + ".txt", objective_time, derivative_time);
+        save_objective_to_file(output_dir + input_basename + "_F_" + module_basename + ".txt", objective_time);
+        save_gradient_to_file(output_dir + input_basename + "_J_" + module_basename + ".txt", gradient_size, output.gradient.data());
```

Comment:
This code expects `output_dir` to have a trailing slash. We need to ensure, it's there.

Reply:
Or just call the variable ``output_prefix``, then if it ends in a slash, it will name a directory, otherwise just a filename prefix.